lib/mp-readline/readline.c: Added backward/forward/kill-word support.

derekenos opened the pull request:
This PR adds forward-word (Alt-f), backward-word (Alt-b), kill-word (Alt-d), and backward-kill-word (Alt-Backspace) support to readline.

derekenos committed:
lib/mp-readline/readline.c: Added backward/forward/kill-word support.

stinos commented:
This is a good addition in general I think as it makes the REPL more useful, problem for me personally is that I don't use standard readline shortcuts for this but the 'gui text editor-style' ones (Ctrl+Left Arrow, Ctrl+Del etc), which as far as I know this is not that uncommon (CPython's REPL also has a subset of it for instance). Now the core logic is the same, so if you refactor what is now in the switch statement into functions or macros it would be fairly easy to provide both sets of shortcuts. possibly configurable?

derekenos commented:
I'll be happy to add hooks for `Ctrl-Left` and `Ctrl-Right` which I see are implemented in the CPython REPL, though it'll take more work than I initially thought because these escape sequences comprise more characters than the current logic supports.

`Ctrl-Del`, however, doesn't appear to be supported in the CPython REPL; when I type this, I just see:`5~`

I'm also getting some ctypes-related build errors for various ports that I don't really understand and need to investigate.

dpgeorge commented on lib/mp-readline/readline.c:
```diff
@@ -98,6 +99,38 @@ typedef struct _readline_t {
+  size_t cursor_pos = rl.cursor_pos;
+  bool in_leading_nonalphanum = true;
+  while (cursor_pos > 0) {
+    if (!isalnum((unsigned char) rl.line->buf[cursor_pos - 1])) {
```
Please see `py/unicode.c` for functions that can be used to test character sets.  We can't rely on ctype.h.

derekenos replied:
Now using `unichar_is*` funcs: https://github.com/micropython/micropython/pull/5024/commits/00b2a0e2609ed84df5be05af728c551c457e622f#diff-ec323dc5c7454143942798519678d424R113

dpgeorge commented on lib/mp-readline/readline.c:
```diff
@@ -98,6 +99,38 @@ typedef struct _readline_t {
 
 STATIC readline_t rl;
 
+size_t get_previous_word_cursor_pos() {
```
C requires `(void)` for functions without args.

derekenos replied:
I haven't written C in a very long time and would like to believe that if it compiles it's correct :laughing:

dhylands replied:
In C, `void foo()` says that the function foo can take any number of arguments, where `void foo(void)` says that the function foo takes no arguments. Both are syntactically correct but have  different semantic interpretations.

dpgeorge commented on lib/mp-readline/readline.c:
```diff
@@ -98,6 +99,38 @@ typedef struct _readline_t {
+  return cursor_pos;
+}
+
+size_t get_next_word_cursor_pos() {
```
Would be good to combine this function with the one above, parameterised by a "direction" argument, to reduce code size.

derekenos replied:
Combined into single [`size_t get_word_cursor_pos(bool direction)`](https://github.com/micropython/micropython/pull/5024/commits/00b2a0e2609ed84df5be05af728c551c457e622f#diff-ec323dc5c7454143942798519678d424R105)

dpgeorge commented on lib/mp-readline/readline.c:
```diff
@@ -214,10 +247,40 @@ int readline_process_char(int c) {
             redraw_step_forward = 1;
         }
     } else if (rl.escape_seq == ESEQ_ESC) {
+        size_t num_chars;
```
This should be a local variable, local in scope to the case's that use it.

derekenos replied:
I had initially tried to do this but the compiler was complaining at me, but have since figured out that I need to wrap the `case` body in `{ ... }`.

dpgeorge reviewed (commented):
Thanks for the contribution.  Please note that MicroPython needs to maintain a relatively small footprint, so adding functionality like this is not always the right way to go.  Nevertheless, if it's a small impact in code size for a large impact in usability then it'd be considered for inclusion.

It might be good to conditionally include the new features here with the existing `MICROPY_REPL_EMACS_KEYS` option.

stinos commented on lib/mp-readline/readline.c:
```diff
@@ -98,6 +99,38 @@ typedef struct _readline_t {
+  size_t cursor_pos = rl.cursor_pos;
+  bool in_leading_nonalphanum = true;
+  while (cursor_pos > 0) {
+    if (!isalnum((unsigned char) rl.line->buf[cursor_pos - 1])) {
```
Ah and indent should be 4 tabs, see surrounding code

derekenos replied:
Fixed.

stinos commented:
> Ctrl-Del, however, doesn't appear to be supported in the CPython REPL; when I type this, I just see:5~

Yes I know, quite an oversight imo. Which is likely why it gets usually reimplemented in the IDEs using it. Anyway I can understand if this is considered too non-standard for a REPL but I can always implement it in a fork.

derekenos committed:
lib/mp-readline/readline.c: Consolidated prev/next word funcs. Stopped using ctypes. Fixed formatting.

derekenos commented:
Thank you, @dpgeorge, for this project! I've done a lot of microcontroller firmware development, some in C but mostly Assembly, and the ability to use Python on these platforms is really amazing.

I understand that MicroPython needs to stay small, but these key bindings will make the REPL more functional for me, and hopefully others.

I apologize in advance if my code is doing anything that's obviously ridiculous. I'll look into making this feature condition with that `MICROPY_REPL_EMACS_KEYS`

derekenos committed:
lib/mp-readline/readline.c: Made new features conditional on MICROPY_REPL_EMACS_KEYS

derekenos committed:
lib/mp-readline/readline.h: Added missing get_word_cursor_pos() prototype.

dpgeorge commented on lib/mp-readline/readline.c:
```diff
@@ -98,6 +102,38 @@ typedef struct _readline_t {
+#define FORWARD true
+#define BACKWARD false
+
+size_t get_word_cursor_pos(bool direction) {
```
This can be made a `STATIC` function, since it's local to this file.

Also `direction` can be an `int` taking value 1 or -1, then there's no need to convert it to `cursor_pos_step`, nor have `FORWARD` and `BACKWARD` constants.

dpgeorge commented on lib/mp-readline/readline.c:
```diff
@@ -98,6 +102,38 @@ typedef struct _readline_t {
+    char buf_idx_offset = direction == FORWARD ? 0 : -1;
+    size_t cursor_pos = rl.cursor_pos;
+    bool in_leading_nonalphanum = true;
+    char c;
```
`char c` should go below in the `while` loop.

dpgeorge commented on lib/mp-readline/readline.c:
```diff
@@ -98,6 +102,38 @@ typedef struct _readline_t {
+            if (!in_leading_nonalphanum) {
+                break;
+            }
+        } else if (in_leading_nonalphanum) {
```
This check is not needed, can just be an `else` (saves code size).

dpgeorge commented on lib/mp-readline/readline.c:
```diff
@@ -98,6 +102,38 @@ typedef struct _readline_t {
+        } else if (in_leading_nonalphanum) {
+            in_leading_nonalphanum = false;
+        }
+        if (direction == FORWARD) {
```
No need for this check, just unconditionally test if `cursor_pos >= rl.line->len`

dpgeorge commented on lib/mp-readline/readline.c:
```diff
@@ -98,6 +102,38 @@ typedef struct _readline_t {
+        }
+        cursor_pos += cursor_pos_step;
+    }
+    return cursor_pos;
```
Can probably return `cursor_pos - rl.cursor_pos` here, since all callers just want the number of chars skipped.

dpgeorge commented:
Eventually this would also need tests, see eg `tests/cmdline/repl_emacs_keys.py`

jimmo commented:
I know `M-del` is the Emacs shortcut for `backward-kill-word`, but GNU readline also supports `C-w` which is a far easier option to type. So if you're adding the implementation anyway, would you be prepared to spend a few extra bytes on making `C-w` also trigger `backward-kill-word` ?

stinos commented:
That makes already for 3 different shortcuts people want for this. So whether or not they all get implemented in the main port, I really think it's for the best if all functionality is in seperate functions and/or macros so people who want a shortcut not available in the main port can easily add it in their fork by referring to said functions/macros simply by adding switch cases and populating them with clls like `word_backward` `kill_word_forward` etc.

dpgeorge commented:
Since alt-<x> shortcuts are not currently used anywhere in this readline implementation it would be important to check -- for multiple terminal programs -- that this PR uses the correct escape sequences for it.  For me (Linux + urxvt) it works correctly.

> That makes already for 3 different shortcuts people want for this.

The simplest option is not to add any :)

Adding a way to customise the shortcuts via a config file would add more code than just supporting all the different keys by default.

So I would suggest that by default there just be one shortcut and it be consistent with some existing application, eg emacs.

stinos commented:
Doesn't work on windows port but that's because the mapping in windows_mphal.c doesn't implement `Alt` so that's a different issue. Works in e.g. xfce4-terminal and Terminator but not in xterm here.

derekenos committed:
lib/mp-readline/readline.c: Refactored get_word_cursor_pos().

Per: https://github.com/micropython/micropython/pull/5024#discussion_r316939927

Made the function STATIC, changed direction to an int, moved 'char c' into the
while loop, removed unnecessary check, renamed function to
get_next_word_cursor_offset() and made it return the signed cursor position offset
to the next word in the specified direction.

Also changed the shortcut comments to reflect the Emacs key notation.

Jongy commented:
@derekenos do you plan on finishing this? I wrote a quick `forward-word` and `backward-word` now and I thought to check in the existing PRs only after I was done... ;(

I can pick it up / continue with my impl. I've also made it work for "Ctrl+Right" / "Ctrl+Left" since I usually use them.

derekenos commented:
@Jongy please feel free to pick up this up or continue with your own. I was at the point of adding tests but became busy with other things. I considered closing this PR but hoped to come back to it, and didn't want the feature-relevant conversation contained herein to disappear from sight.

Jongy commented:
#5420 was merged.
